### ISSUE TITLE
`einsum` is ~20X slower than manually multiplying and summing

### ISSUE THREAD

**Issue by SamPruden** [pytorch/pytorch#32591]
## 🐛 Bug
A manual multiplication and summation `(a * b).sum(dim = (-3, -2, -1))` is about 20X faster than the equivalent `einsum`.

## To Reproduce

```Python3
a = torch.rand(1, 1, 16, 2, 16, 2, 16, 2, 2, 2, 2, device = "cuda")
b = torch.rand(729, 1, 1, 2, 1, 2, 1, 2, 2, 2, 2, device = "cuda")

# Warmup
for i in range(100): output1 = (a * b).sum(dim = (-3, -2, -1))

# Method 1
with torch.autograd.profiler.profile(use_cuda = True) as prof:
  output1 = (a * b).sum(dim = (-3, -2, -1))
print(prof.key_averages().table(sort_by="cuda_time_total"))

# Warmup
for i in range(100):
  (a2, b2) = torch.broadcast_tensors(a, b)
  output2 = torch.einsum("...ijk, ...ijk -> ...", a2, b2)

# Method 2
with torch.autograd.profiler.profile(use_cuda = True) as prof:
  (a2, b2) = torch.broadcast_tensors(a, b)
  output2 = torch.einsum("...ijk, ...ijk -> ...", a2, b2)
print(prof.key_averages().table(sort_by="cuda_time_total"))
```
```
-------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
Name     Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     CUDA total %     CUDA total       CUDA time avg    Number of Calls  
-------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
sum      28.93%           43.848us         28.93%           43.848us         43.848us         65.81%           8.978ms          8.978ms          1                
mul      71.07%           107.713us        71.07%           107.713us        107.713us        34.19%           4.665ms          4.665ms          1                
-------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
Self CPU time total: 151.561us
CUDA time total: 13.644ms

---------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
Name                   Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     CUDA total %     CUDA total       CUDA time avg    Number of Calls  
---------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
einsum                 4.05%            59.449us         97.88%           1.435ms          1.435ms          48.62%           123.798ms        123.798ms        1                
bmm                    85.15%           1.248ms          85.15%           1.248ms          1.248ms          45.83%           116.694ms        116.694ms        1                
reshape                1.35%            19.770us         6.36%            93.237us         46.619us         2.76%            7.031ms          3.516ms          2                
clone                  3.96%            58.021us         3.96%            58.021us         29.011us         2.75%            7.012ms          3.506ms          2                
broadcast_tensors      0.79%            11.624us         2.12%            31.095us         31.095us         0.01%            31.008us         31.008us         1                
expand                 1.33%            19.471us         1.33%            19.471us         9.736us          0.01%            20.480us         10.240us         2                
permute                1.54%            22.535us         1.54%            22.535us         4.507us          0.01%            19.327us         3.865us          5                
view                   0.78%            11.442us         0.78%            11.442us         5.721us          0.00%            4.094us          2.047us          2                
_unsafe_view           1.05%            15.446us         1.05%            15.446us         7.723us          0.00%            3.968us          1.984us          2                
---------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
Self CPU time total: 1.466ms
CUDA time total: 254.614ms
```

## Expected behavior

`einsum` to be no slower than manual. I actually expected / hoped for it to be faster as it avoids creating the large intermediary result of the multiplication.

## Environment

The environment is Google Colab with a fresh GPU instance on which `!pip install -q torch==1.4.0 torchvision==0.5.0` has been run as Colab is still on `1.3.1` by default.

```
PyTorch version: 1.4.0
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: version 3.12.0

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: GPU 0: Tesla P100-PCIE-16GB
Nvidia driver version: 418.67
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5

Versions of relevant libraries:
[pip3] numpy==1.17.5
[pip3] torch==1.4.0
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.3.1
[pip3] torchvision==0.5.0
[conda] Could not collect
```

cc @ngimel @vincentqb @vishwakftw @jianyuh @nikitaved @pearu @VitalyFedyunin @mruberry

**Comment by ngimel** [comment]
cc @t-vi

**Comment by ngimel** [comment]
Thank you for reporting this issue. They way einsum is currently implemented there is no guarantee that it will be faster than matmul (in fact, it goes through batch matmul most of the time). If you want to improve performance by avoiding materialization of large intermediate tensors keops may be of interest. #28341 also might be related.

**Comment by t-vi** [comment]
This seems to be a limitation of the strategy to reduce to matmuls. Maybe one should revisit a TensorIterator-based implementation for GPU. I tried this a while ago in some branch but in the end didn't pursue it because I was more interested in the CPU case at that point.

**Comment by yaroslavvb** [comment]
This would be a good building block to support
```
BLAS0=broadcasted pointwise multiplication of two tensors followed by sum reduction
```

You could then compute larger einsums in an optimal way by using the schedule from `opt_einsum`, which is a sequence of GEMM/DOT/BLAS0 calls

```
import opt_einsum as oe
import numpy as np
print(oe.contract_path('01,12,23,34,45,56->03', *[np.ones((20,20))]*6, optimize='dp'))
```

**Comment by ngimel** [comment]
@yaroslavvb if pytorch implemented this primitive (broadcasted pointwise multiplication of two tensors followed by sum reduction), will opt_einsum then need to be taught to call it in appropriate circumstances?

**Comment by SamPruden** [comment]
Thanks for pointing out keops @ngimel, I hadn't heard of that before. That looks like a very useful tool, although having played around with it a bit I don't think it's going to give a performance boost for my use case here. I'll certainly find uses for it in the future.

> BLAS0=broadcasted pointwise multiplication of two tensors followed by sum reduction

That would be a very useful function to have. I actually found this issue after posting [this question](https://discuss.pytorch.org/t/fast-batched-multiplication-and-sum-over-last-3-dimensions-on-arbitrary-view-without-reshape/67564) asking for such a function as I assumed it must already exist.

My use case is a custom convolution like operation, so something that performs well when chained with `Tensor.Unfold` would be very helpful. So far I've found that the best performance for a custom 3D convolution comes from the `(a * b).sum(dir = (-3, -2, -1))` implementation, although for the size I'm working with it's actually proving faster to use `conv3d` with a kernel that's bigger than it needs to be and mostly full of zeros, so that's what I'm using at the moment.

I think that this convolution use case is worth thinking about because it means a good implementation should be optimised for fairly small reductions. In my case I have a lot of 2X2X2 patches to be multiplied and summed. Perhaps an implementation should branch by shape and use a different method for large reductions which should themselves be parallel, vs small reductions which can happen in a single thread. I don't know if that type of thinking applies to how these types of algorithms are implemented here, but that's my first thought as somebody with GPU experience on the graphics side of things.

**Comment by yaroslavvb** [comment]
@ngimel I think it's easier to get contraction order from opt_einsum and do the rest in PyTorch

Here's an example of how to turn contraction order into a sequence of einsum calls, these map directly to optimized primitives like BLAS0 or GEMM

https://colab.research.google.com/drive/1ItfFMp6WGdZLSrFtI-ppnVcD2ahLxIHg

**Comment by ngimel** [comment]
Cool! @yaroslavvb can you file an issue with this feature request?

**Comment by fmassa** [comment]
> BLAS0=broadcasted pointwise multiplication of two tensors followed by sum reduction

I believe we might have something similar (or almost) implemented, see 

https://github.com/pytorch/pytorch/blob/e7fe64f6a65cd427e503491f192c14476e18033b/aten/src/ATen/native/Linear.cpp#L33

**Comment by ngimel** [comment]
@fmassa it's likely already used by einsum. But apparently there are cases when going through bmm, as that function does, is pretty inefficient. BTW, nvidia provides a cutensor library that should have efficient implementations for tensor contraction ops, I wonder if it would help https://developer.nvidia.com/cutensor

**Comment by yaroslavvb** [comment]
@fmassa indeed, that seems to be it. Is this what's getting called in the instance above? It seems slower than pointwise multiply followed by reduce. A simpler example with 20x slowdown

```
a = torch.rand(2000000, 10, device = "cuda")
b = torch.rand(2000000, 10, device = "cuda")
output1 = (a * b).sum(dim=1)                 # 2 ms
output2 = torch.einsum("ij,ij->i", a, b)     # 40 ms
```

**Comment by ngimel** [comment]
@yaroslavvb looks like it. Since batch size for `bmm` is 2000000 in this case, and cublas has an internal limitation of 65K for the batch size, batched gemv kernel is called 31 times, and kernel itself looks to be not particularly efficient.

**Comment by yaroslavvb** [comment]
@ngimel `cutensorContraction` function in that library is essentially a drop-in replacement of einsum (except perhaps 'ii->i' case), seems pretty promising!

**Comment by bkmi** [comment]
I was looking into cutensor and thought that it does look promising. The only problem is that it is [limited to very powerful gpus](https://docs.nvidia.com/cuda/cutensor/user_guide.html#supported-gpus), i.e. compute >= 7.0.

I have not tried using cutensor yet so I don't know if that is a strict requirement or a reccomendation.

**Comment by heitorschueroff** [comment]
Reopening this issue for further investigation. I ran some benchmarks to diagnose in which conditions is multiplying and summing faster than einsum. What I found is that if the batch dimensions match in size, then for most cases multiplying and summing will be faster on CUDA except for very large batch AND input size. However, when the batch dimensions are 1 for one of the inputs, then einsum is faster. The reason I believe is that in the first case einsum would call bmm with shapes b x 1 x n and b x n x 1, so all the optimizations with blocking in bmm would not really benefit this case. But, when one of the inputs have batch size 1, einsum will actually move that dimension out of the batch size and into the added tensor dimension for the matrix multiplication. For instance, the case with [1000, 1, 10] x [1, 1000, 10] calls into bmm with shapes [1, 1000, 10] x [1, 10, 1000]. Following are the benchmarks.

## Benchmark script
```python
import torch
from torch.utils import benchmark

results = []
for b in [10, 10000, 2000000]:
    for n in [10, 100, 10000, 1000000]:
        if b * n >= 1000000000:
            continue

        description = f'[{b}, {n}]'

        x = torch.rand(b, n, device='cuda')
        y = torch.rand(b, n, device='cuda')

        results.append(benchmark.Timer(
            stmt='(x * y).sum(dim=-1)',
            globals={'x': x, 'y': y},
            description=description,
        ).blocked_autorange())

        results.append(benchmark.Timer(
            stmt='torch.einsum("...j,...j->...", x, y)',
            globals={'x': x, 'y': y},
            description=description,
        ).blocked_autorange())

compare = benchmark.Compare(results)
compare.trim_significant_figures()
compare.colorize()
compare.print()


results = []
for b in [10, 100, 1000]:
    for n in [10, 100, 10000, 1000000]:
        if b * b * n >= 1000000000:
            continue

        description = f'[{b}, {b}, {n}]'

        x = torch.rand(b, b, n, device='cuda')
        y = torch.rand(b, b, n, device='cuda')

        results.append(benchmark.Timer(
            stmt='(x * y).sum(dim=-1)',
            globals={'x': x, 'y': y},
            description=description,
        ).blocked_autorange())

        results.append(benchmark.Timer(
            stmt='torch.einsum("...j,...j->...", x, y)',
            globals={'x': x, 'y': y},
            description=description,
        ).blocked_autorange())

compare = benchmark.Compare(results)
compare.trim_significant_figures()
compare.colorize()
compare.print()


results = []
for b in [10, 100, 1000]:
    for n in [10, 100, 10000, 1000000]:
        if b * b * n >= 1000000000:
            continue

        description = f'[{b}, {b}, {n}]'

        x = torch.rand(b, 1, n, device='cuda')
        y = torch.rand(1, b, n, device='cuda')

        results.append(benchmark.Timer(
            stmt='(x * y).sum(dim=-1)',
            globals={'x': x, 'y': y},
            description=description,
        ).blocked_autorange())

        results.append(benchmark.Timer(
            stmt='torch.einsum("...j,...j->...", x, y)',
            globals={'x': x, 'y': y},
            description=description,
        ).blocked_autorange())

compare = benchmark.Compare(results)
compare.trim_significant_figures()
compare.colorize()
compare.print()
```

## Benchmark results
```
[----------------------------------------------------------------------------------------------  ----------------------------------------------------------------------------------------------]
                                            |  [10, 10]  |  [10, 100]  |  [10, 10000]  |  [10, 1000000]  |  [10000, 10]  |  [10000, 100]  |  [10000, 10000]  |  [2000000, 10]  |  [2000000, 100]
1 threads: -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
      (x * y).sum(dim=-1)                   |     21     |      22     |       21      |        314      |       21      |      44.9      |       2837       |       1140      |       7897     
      torch.einsum("...j,...j->...", x, y)  |     60     |      40     |       43      |       6220      |       40      |      54.3      |       1910       |       8328      |      10240     

Times are in microseconds (us).

[--------------------------------------------------------------------------------------------------------------  -------------------------------------------------------------------------------------------------------------]
                                            |  [10, 10, 10]  |  [10, 10, 100]  |  [10, 10, 10000]  |  [10, 10, 1000000]  |  [100, 100, 10]  |  [100, 100, 100]  |  [100, 100, 10000]  |  [1000, 1000, 10]  |  [1000, 1000, 100]
1 threads: --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
      (x * y).sum(dim=-1)                   |       21       |        21       |        33.8       |         2860        |        21        |        45.0       |         2837        |         572        |         3949      
      torch.einsum("...j,...j->...", x, y)  |       40       |        40       |        76.4       |         7300        |        41        |        54.3       |         1909        |        4174        |         5125      

Times are in microseconds (us).

[--------------------------------------------------------------------------------------------------------------  -------------------------------------------------------------------------------------------------------------]
                                            |  [10, 10, 10]  |  [10, 10, 100]  |  [10, 10, 10000]  |  [10, 10, 1000000]  |  [100, 100, 10]  |  [100, 100, 100]  |  [100, 100, 10000]  |  [1000, 1000, 10]  |  [1000, 1000, 100]
1 threads: --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
      (x * y).sum(dim=-1)                   |       23       |        23       |        38.0       |         3219        |        22        |        49.2       |         3209        |        620         |         4314      
      torch.einsum("...j,...j->...", x, y)  |       42       |        41       |        68.3       |         3812        |        41        |        41.6       |           95        |         56         |           41      

Times are in microseconds (us).
```

**Comment by heitorschueroff** [comment]
PR https://github.com/pytorch/pytorch/pull/48184 fixes the slowdown by choosing between multiply/sum and bmm depending on the input size however it introduces regression for other functions that depend on sumproduct_pair. A solution requires further design.

**Comment by Linux-cpp-lisp** [comment]
Are there any updates on cuTENSOR in PyTorch? (For example, can it be used as a backend for `torch.tensordot`, which seems to be the exact same computation as https://docs.nvidia.com/cuda/cutensor/api/cutensor.html#cutensorcontraction?)

**Comment by heitorschueroff** [comment]
Current implementation of `torch.einsum` fixed this performance problem as shown in the following benchmarks, where mul/sum refers to the code `(a * b).sum(dim=(-3, -2, -1))`.

### CPU Benchmark
```
[---------------------------------  ---------------------------------]
                          |  mul/sum  |  torch.einsum  |  numpy.einsum
1 threads: -----------------------------------------------------------
      ...ijk,...ijk->...  |   14080   |      65.6      |      2336    

Times are in milliseconds (ms).
```

### CUDA Benchmark
```
[---------------------------------  ---------------------------------]
                          |  mul/sum  |  torch.einsum  |  numpy.einsum
1 threads: -----------------------------------------------------------
      ...ijk,...ijk->...  |    4520   |     585.7      |    3321000   

Times are in microseconds (us).
```
